edge/pkg/clusterd/mission_deployer.go

```diff
@@ -0,0 +1,348 @@
+	if err != nil {
+		// log the error and move on to apply the mission content
+		klog.Errorf("Error in applying mission CRD: %v. Moving on.", err)
+	} else {
+		ddeployMissionCmd := fmt.Sprintf("printf \"%s\" | %s apply --kubeconfig=%s -f - ", missionYaml, config.Config.KubectlCli, config.Config.Kubeconfig)
+		output, err := util.ExecCommandLine(ddeployMissionCmd)
+		if err != nil {
+			klog.Errorf("Failed to apply mission %v: %v", mission.Name, err)
+		} else {
+			if strings.Contains(output, "created") {
+				klog.Infof("Mission %v is created ", mission.Name)
+			} else {
+				klog.V(4).Infof("Mission %v is configured.", mission.Name)
+			}
+		}
+	}
+
+	if !m.isMatchingMission(mission) {
+		klog.V(3).Infof("Mission %v does not match this cluster, skip the content applying", mission.Name)
+	} else {
+		if strings.TrimSpace(mission.Spec.Content) != "" {
+			deployContentCmd := fmt.Sprintf("printf \"%s\" | %s apply --kubeconfig=%s -f - ", mission.Spec.Content, config.Config.KubectlCli, config.Config.Kubeconfig)
+			output, err := util.ExecCommandLine(deployContentCmd)
+			if err != nil {
+				klog.Errorf("Failed to apply the content of mission %v: %v", mission.Name, err)
+			} else {
+				if strings.Contains(output, "unchanged") {
+					klog.V(4).Infof("The content of mission %v is unchanged ", mission.Name)
+				} else {
+					klog.Infof("The content of mission %v applied successfully ", mission.Name)
+				}
+			}
+		}
+	}
+
+	m.UpdateState(mission, false)
+
+	return nil
+}
+
+func (m *MissionDeployer) DeleteMission(mission *edgeclustersv1.Mission) error {
+	cacheLock.Lock()
+	delete(m.MissionMatch, mission.Name)
+	cacheLock.Unlock()
+	if !m.isMatchingMission(mission) {
+		klog.V(4).Infof("Mission %v does not match this cluster", mission.Name)
+	} else {
+		if strings.TrimSpace(mission.Spec.Content) != "" {
+			deployContentCmd := fmt.Sprintf("printf \"%s\" | %s delete --kubeconfig=%s -f - ", mission.Spec.Content, config.Config.KubectlCli, config.Config.Kubeconfig)
+			_, err := util.ExecCommandLine(deployContentCmd)
+			if err != nil {
+				klog.Errorf("Failed to revert the content of mission %v: %v", mission.Name, err)
+			} else {
+				klog.Infof("The content of mission %v is reverted.", mission.Name)
+			}
+		}
+	}
+
+	deleteMissionCmd := fmt.Sprintf("%s delete mission %s --kubeconfig=%s", config.Config.KubectlCli, mission.Name, config.Config.Kubeconfig)
+	if _, err := util.ExecCommandLine(deleteMissionCmd); err != nil {
+		return fmt.Errorf("Failed to delete mission %v: %v", mission.Name, err)
+	}
+
+	klog.Infof("Mission %v deleted successfully ", mission.Name)
+
+	return nil
+}
+
+func (m *MissionDeployer) DeleteMissionByName(name string) error {
+	mission, err := helper.GetMissionByName(name)
+	if err != nil {
+		return err
+	}
+
+	return m.DeleteMission(mission)
+}
+
+func (m *MissionDeployer) isMatchingMission(mission *edgeclustersv1.Mission) bool {
+	// if the placement field is empty, it matches all the edge clusters
```

Comment:
discussion: a wild card should be used to match all the edge cluster, and an an empty field should match none.

also, in future, access control should be used so a user wouldn't be able to deploy to "all" the clusters, e.g. those that the user doesn't own. 

let's have a separate design on placement someday.

Reply:
Yep, we need a separate design on this. 

Currently, we are re-using the constraint expression in Kubernentes federation, which may not be applicable to the situation of edge clsuters.

issue created to track this: https://github.com/CentaurusInfra/fornax/issues/26